File: frontend/src/types/lab.ts
// Laboratory module types — mirrors the backend schemas/lab.py shapes.

export interface LabTestParameterTemplate {
  name: string;
  unit?: string | null;
  reference_range?: string | null;
  sequence?: number | null;
  section?: string | null;
}

export interface LabTest {
  id: string;
  hospital_id: string;
  name: string;
  code: string;
  category?: string | null;
  sample_type?: string | null;
  price: number;
  unit?: string | null;
  reference_range?: string | null;
  turnaround_hours?: number | null;
  is_active: boolean;
  report_template?: LabTestParameterTemplate[] | null;
  created_at: string;
  updated_at: string;
}

export interface LabTestCreateData {
  name: string;
  code: string;
  category?: string;
  sample_type?: string;
  price: number;
  unit?: string;
  reference_range?: string;
  turnaround_hours?: number;
  is_active?: boolean;
  report_template?: LabTestParameterTemplate[];
}

export interface LabTestListResponse {
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  data: LabTest[];
}

export type LabResultFlag = 'normal' | 'high' | 'low' | 'abnormal';
export type LabItemStatus = 'ordered' | 'sample_collected' | 'in_progress' | 'completed' | 'cancelled';
export type LabQueueStatus = 'waiting' | 'being_served' | 'collected';
export type LabReportStatus = 'pending' | 'completed' | 'finalized';

export interface LabResultParameter {
  name: string;
  value: string;
  unit?: string | null;
  reference_range?: string | null;
  flag?: LabResultFlag | null;
  sequence?: number | null;
  section?: string | null;
}

export interface LabOrderItem {
  id: string;
  lab_order_id: string;
  lab_test_id: string;
  test_name: string;
  price: number;
  status: LabItemStatus;
  parameters: LabResultParameter[];
  report_template: LabTestParameterTemplate[];
  // Legacy single-value fields — read-only fallback for rows entered before
  // structured `parameters` existed.
  result_value?: string | null;
  result_unit?: string | null;
  reference_range?: string | null;
  result_flag?: LabResultFlag | null;
  result_notes?: string | null;
  resulted_at?: string | null;
  resulted_by?: string | null;
  created_at: string;
}

export interface LabOrder {
  id: string;
  hospital_id: string;
  order_number: string;
  patient_id: string;
  doctor_id: string | null;
  appointment_id?: string | null;
  prescription_id?: string | null;
  notes?: string | null;
  is_finalized: boolean;
  status: string;
  report_status: LabReportStatus;
  finalized_at?: string | null;
  queue_token?: number | null;
  queue_status?: LabQueueStatus | null;
  queue_called_at?: string | null;
  created_at: string;
  updated_at?: string | null;
  // Enriched
  patient_name?: string | null;
  doctor_name?: string | null;
  finalized_by_name?: string | null;
  total_amount?: number | null;
  payment_status?: string | null;
  sale_id?: string | null;
  items: LabOrderItem[];
}

export interface LabOrderCreateData {
  patient_id: string;
  doctor_id?: string;
  appointment_id?: string;
  prescription_id?: string;
  test_ids: string[];
  notes?: string;
}

export interface LabResultEntryData {
  parameters: LabResultParameter[];
  result_notes?: string;
}

export interface LabQueueEntry {
  id: string;
  order_number: string;
  patient_name?: string | null;
  queue_token?: number | null;
  queue_status: LabQueueStatus;
  status: string;
  total_amount: number;
  payment_status: string;
  created_at: string;
  queue_called_at?: string | null;
}

export interface LabSale {
  id: string;
  hospital_id: string;
  sale_number: string;
  lab_order_id: string;
  patient_id: string;
  invoice_id?: string | null;
  subtotal: number;
  discount_amount: number;
  tax_amount: number;
  total_amount: number;
  payment_method: string;
  payment_status: string;
  amount_tendered: number;
  advance_amount: number;
  paid_amount: number;
  balance_amount: number;
  status: string;
  created_at: string;
  updated_at?: string | null;
}

// Row shape for the standalone Lab Billing worklist (LabBilling.tsx) —
// mirrors backend schemas.lab.LabBillingItemResponse.
export interface LabBillingItem {
  id: string;
  order_number: string;
  patient_id: string;
  patient_name?: string | null;
  total_amount: number;
  paid_amount: number;
  balance_amount: number;
  payment_status: string;
  report_status: LabReportStatus;
  created_at: string;
}

export interface LabBillingListResponse {
  total: number;
  page: number;
  limit: number;
  total_pages: number;
  data: LabBillingItem[];
}

export interface LabDashboard {
  total_tests: number;
  today_orders_count: number;
  waiting_count: number;
  pending_results_count: number;
  today_revenue: number;
}

export interface PatientLabResultItem {
  id: string;
  test_name: string;
  status: LabItemStatus;
  parameters: LabResultParameter[];
  result_notes?: string | null;
  resulted_at?: string | null;
}

export interface PatientLabResult {
  id: string;
  order_number: string;
  status: string;
  created_at: string;
  doctor_name?: string | null;
  finalized_at?: string | null;
  finalized_by_name?: string | null;
  items: PatientLabResultItem[];
}

// External referral letter (e.g. to a consultant radiologist) — printed and
// filled by lab staff, independent of the LabOrder/LabTest ordering flow.
export interface LabReferral {
  id: string;
  hospital_id: string;
  referral_number: string;
  patient_id: string;
  recipient_title: string;
  recipient_location?: string | null;
  case_details?: string | null;
  investigation: string;
  remarks?: string | null;
  referring_doctor_name: string;
  created_at: string;
  patient_name?: string | null;
}

export interface LabReferralCreateData {
  patient_id: string;
  recipient_title: string;
  recipient_location?: string;
  case_details?: string;
  investigation: string;
  remarks?: string;
  referring_doctor_name: string;
}

File: backend/app/schemas/lab.py
"""
Laboratory module Pydantic schemas — test catalog, orders (+ result rows), sales.

Follows schemas/optical.py's shape: `_orm_to_dict` transform on responses,
denormalized/enriched fields set by the service layer.
"""
from pydantic import BaseModel, Field, ConfigDict, model_validator
from typing import Optional, Any
from datetime import datetime
from decimal import Decimal


def _orm_to_dict(data: Any) -> Any:
    """Convert a SQLAlchemy ORM instance to a dict, stringifying UUIDs."""
    if hasattr(data, "__mapper__"):
        d = {}
        for attr in data.__mapper__.column_attrs:
            key = attr.key
            val = getattr(data, key)
            if hasattr(val, "hex"):
                val = str(val)
            d[key] = val
        return d
    return data


# ══════════════════════════════════════════════════
# Lab Test (catalog)
# ══════════════════════════════════════════════════
class LabTestParameterTemplate(BaseModel):
    """One row of a test's structured report layout — e.g. a CBC's
    individual parameters (Hemoglobin, WBC, ...). Populated once real report
    templates are supplied; result entry falls back to a single free-text
    row when a test has none."""
    name: str = Field(..., min_length=1, max_length=200)
    unit: Optional[str] = Field(None, max_length=30)
    reference_range: Optional[str] = Field(None, max_length=200)
    section: Optional[str] = Field(None, max_length=100)
    sequence: Optional[int] = None


class LabTestCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    code: str = Field(..., min_length=1, max_length=30)
    category: Optional[str] = Field(None, max_length=100)
    sample_type: Optional[str] = Field(None, max_length=50)
    price: Decimal = Field(default=Decimal("0"), ge=0)
    unit: Optional[str] = Field(None, max_length=30)
    reference_range: Optional[str] = Field(None, max_length=200)
    turnaround_hours: Optional[int] = Field(None, ge=0)
    is_active: bool = True
    report_template: Optional[list[LabTestParameterTemplate]] = None


class LabTestUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    code: Optional[str] = Field(None, min_length=1, max_length=30)
    category: Optional[str] = Field(None, max_length=100)
    sample_type: Optional[str] = Field(None, max_length=50)
    price: Optional[Decimal] = Field(None, ge=0)
    unit: Optional[str] = Field(None, max_length=30)
    reference_range: Optional[str] = Field(None, max_length=200)
    turnaround_hours: Optional[int] = Field(None, ge=0)
    is_active: Optional[bool] = None
    report_template: Optional[list[LabTestParameterTemplate]] = None


class LabTestResponse(BaseModel):
    id: str
    hospital_id: str
    name: str
    code: str
    category: Optional[str] = None
    sample_type: Optional[str] = None
    price: Decimal
    unit: Optional[str] = None
    reference_range: Optional[str] = None
    turnaround_hours: Optional[int] = None
    is_active: bool = True
    report_template: Optional[list[LabTestParameterTemplate]] = None
    created_at: datetime
    updated_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class LabTestListResponse(BaseModel):
    total: int
    page: int
    limit: int
    total_pages: int
    data: list[LabTestResponse]


# ══════════════════════════════════════════════════
# Lab Order
# ══════════════════════════════════════════════════
class LabOrderCreate(BaseModel):
    patient_id: str
    # Optional — when omitted, resolved server-side from the logged-in doctor
    # (the Prescription Builder never re-asks the doctor to select themselves).
    doctor_id: Optional[str] = None
    appointment_id: Optional[str] = None
    prescription_id: Optional[str] = None
    test_ids: list[str] = Field(..., min_length=1)
    notes: Optional[str] = None


class LabResultParameter(BaseModel):
    """One structured result row, matching the shape of
    LabTestParameterTemplate but carrying the entered value + flag."""
    name: str = Field(..., min_length=1, max_length=200)
    value: str = Field(..., max_length=200)
    unit: Optional[str] = Field(None, max_length=30)
    reference_range: Optional[str] = Field(None, max_length=200)
    section: Optional[str] = Field(None, max_length=100)
    flag: Optional[str] = Field(None, pattern="^(normal|high|low|abnormal)$")
    sequence: Optional[int] = None


class LabResultEntry(BaseModel):
    parameters: list[LabResultParameter] = Field(..., min_length=1)
    result_notes: Optional[str] = None


class LabOrderItemResponse(BaseModel):
    id: str
    lab_order_id: str
    lab_test_id: str
    test_name: str
    price: Decimal
    status: str = "ordered"
    parameters: list[LabResultParameter] = []
    report_template: list[LabTestParameterTemplate] = []
    # Legacy single-value fields — read-only fallback for rows entered before
    # structured `parameters` existed; new saves no longer write these.
    result_value: Optional[str] = None
    result_unit: Optional[str] = None
    reference_range: Optional[str] = None
    result_flag: Optional[str] = None
    result_notes: Optional[str] = None
    resulted_at: Optional[datetime] = None
    resulted_by: Optional[str] = None
    created_at: datetime

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        d = _orm_to_dict(data)
        if isinstance(d, dict):
            # JSONB columns are nullable — coerce NULL to an empty list so
            # the non-Optional list fields validate.
            if d.get("parameters") is None:
                d["parameters"] = []
            d.setdefault("report_template", [])
        return d

    model_config = ConfigDict(from_attributes=True)


class LabOrderResponse(BaseModel):
    id: str
    hospital_id: str
    order_number: str
    patient_id: str
    doctor_id: str
    appointment_id: Optional[str] = None
    prescription_id: Optional[str] = None
    notes: Optional[str] = None
    is_finalized: bool = False
    status: str = "ordered"
    report_status: str = "pending"
    finalized_at: Optional[datetime] = None
    queue_token: Optional[int] = None
    queue_status: Optional[str] = None
    queue_called_at: Optional[datetime] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    # Enriched
    patient_name: Optional[str] = None
    doctor_name: Optional[str] = None
    finalized_by_name: Optional[str] = None
    total_amount: Optional[Decimal] = None
    payment_status: Optional[str] = None
    sale_id: Optional[str] = None
    items: list[LabOrderItemResponse] = []

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


# ══════════════════════════════════════════════════
# Lab Queue
# ══════════════════════════════════════════════════
class LabQueueEntryResponse(BaseModel):
    id: str
    order_number: str
    patient_name: Optional[str] = None
    queue_token: Optional[int] = None
    queue_status: str = "waiting"
    status: str = "ordered"
    total_amount: Decimal = Decimal("0")
    payment_status: str = "pending"
    created_at: datetime
    queue_called_at: Optional[datetime] = None


class LabQueueStatusUpdate(BaseModel):
    queue_status: str = Field(..., pattern="^(waiting|being_served|collected)$")


# ══════════════════════════════════════════════════
# Lab Sale
# ══════════════════════════════════════════════════
class LabSaleResponse(BaseModel):
    id: str
    hospital_id: str
    sale_number: str
    lab_order_id: str
    patient_id: str
    invoice_id: Optional[str] = None
    subtotal: Decimal = Decimal("0")
    discount_amount: Decimal = Decimal("0")
    tax_amount: Decimal = Decimal("0")
    total_amount: Decimal = Decimal("0")
    payment_method: str = "cash"
    payment_status: str = "pending"
    amount_tendered: Decimal = Decimal("0")
    advance_amount: Decimal = Decimal("0")
    paid_amount: Decimal = Decimal("0")
    balance_amount: Decimal = Decimal("0")
    status: str = "pending"
    created_at: datetime
    updated_at: Optional[datetime] = None

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


class LabMarkPaidRequest(BaseModel):
    amount_paid: Decimal = Field(..., ge=0)
    payment_method: Optional[str] = None


# ══════════════════════════════════════════════════
# Lab Referral (external referral letter, e.g. to a consultant radiologist)
# ══════════════════════════════════════════════════
class LabReferralCreate(BaseModel):
    patient_id: str
    recipient_title: str = Field(..., min_length=1, max_length=200)
    recipient_location: Optional[str] = Field(None, max_length=200)
    case_details: Optional[str] = None
    investigation: str = Field(..., min_length=1, max_length=200)
    remarks: Optional[str] = None
    referring_doctor_name: str = Field(..., min_length=1, max_length=200)


class LabReferralResponse(BaseModel):
    id: str
    hospital_id: str
    referral_number: str
    patient_id: str
    recipient_title: str
    recipient_location: Optional[str] = None
    case_details: Optional[str] = None
    investigation: str
    remarks: Optional[str] = None
    referring_doctor_name: str
    created_at: datetime
    # Enriched
    patient_name: Optional[str] = None

    @model_validator(mode="before")
    @classmethod
    def transform(cls, data: Any) -> Any:
        return _orm_to_dict(data)

    model_config = ConfigDict(from_attributes=True)


# ══════════════════════════════════════════════════
# Dashboard
# ══════════════════════════════════════════════════
class LabDashboard(BaseModel):
    total_tests: int = 0
    today_orders_count: int = 0
    waiting_count: int = 0
    pending_results_count: int = 0
    today_revenue: Decimal = Decimal("0")
